clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -537,51 +586,137 @@ private void updateGroupSubscription(Set<String> topics) {
-                }
-
-                // also clear the assigned partitions since all have been revoked
+                // revoke all partitions
+                revokedPartitions = new HashSet<>(subscriptions.assignedPartitions());
                 subscriptions.assignFromSubscribed(Collections.emptySet());
 
+                firstException.compareAndSet(null, maybeInvokePartitionsRevoked(revokedPartitions));
                 break;
 
             case COOPERATIVE:
+                // only revoke those partitions that are not in the subscription any more.
+                Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+                revokedPartitions = ownedPartitions.stream()
+                    .filter(tp -> !subscriptions.subscription().contains(tp.topic()))
+                    .collect(Collectors.toSet());
+
+                if (!revokedPartitions.isEmpty()) {
+                    ownedPartitions.removeAll(revokedPartitions);
+                    subscriptions.assignFromSubscribed(ownedPartitions);
+
+                    firstException.compareAndSet(null, maybeInvokePartitionsRevoked(revokedPartitions));
+                }
+
                 break;
         }
 
         isLeader = false;
         subscriptions.resetGroupSubscription();
+
+        if (firstException.get() != null) {
+            throw new KafkaException("User rebalance callback throws an error", firstException.get());
+        }
     }
 
+    /**
+     * @throws KafkaException if the rebalance callback throws exception
+     */
+    @Override
+    public void resetGeneration(String errorMessage) {
+        if (subscriptions.partitionsAutoAssigned()) {
+            // lost all partitions if it is based on subscriptions
+            Set<TopicPartition> lostPartitions = new HashSet<>(subscriptions.assignedPartitions());
+            subscriptions.assignFromSubscribed(Collections.emptySet());
+
+            Exception e = maybeInvokePartitionsLost(errorMessage, lostPartitions);
+
+            if (e != null) {
+                throw new KafkaException("User rebalance callback throws an error", e);
+            }
+        }
+
+        super.resetGeneration(errorMessage);
+    }
+
+    /**
+     * @throws KafkaException if the callback throws exception
+     */
     @Override
     public boolean rejoinNeededOrPending() {
         if (!subscriptions.partitionsAutoAssigned())
             return false;
 
         // we need to rejoin if we performed the assignment and metadata has changed
-        if (assignmentSnapshot != null && !assignmentSnapshot.matches(metadataSnapshot))
+        if (assignmentSnapshot != null && !assignmentSnapshot.matches(metadataSnapshot)) {
+            Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+            Set<TopicPartition> lostPartitions = ownedPartitions.stream()
```

Comment:
A high level comment would be helpful here. I think the point is that we may have "lost" partitions because the topic no longer exists? Maybe we should call this `noLongerExistingPartitions` or something like that just to avoid confusion.